HOLD FOR RELEASE - 73619: Regex Feature (#1049)
replicatedhq/replicated-docs: docs/vendor/admin-console-customize-config-screen.md

For more information about the syntax of the Config custom resource manifest, see [Config](../reference/custom-resource-config) in the _Custom Resources_ section.

## About Regular Expression Validation

You can use [RE2 regular expressions](https://github.com/google/re2/wiki/Syntax) (regex) to validate user input for config items, ensuring conformity to certain standards, such as valid email addresses, password complexity rules, IP addresses, and URLs. This prevents users from deploying an application with a verifiably invalid configuration.

You add the `validation`, `regex`, `pattern` and `message` fields to items in the Config custom resource. Validation is supported for `text`, `textarea`, `password` and `file` config item types.

The following example shows a common password complexity rule:

> ```

Reviewer: I think this block just needs `yaml` to define the color-coding

Author: @paigecalvert It's the weirdest thing, but I cannot fix this issue without removing `yaml` from the code block. The problem is that part of the regex turns gray and italicized unless `yaml` is removed. Meanwhile, that doesn't happen anywhere else, and Pavan could not fix it any other way. I think the best thing to do is for you to approve it again so that Pavan can merge it upon release. Then we can play with it more in a different PR. Would you mind approving it again?

Author: @pavansokkenagaraj The `#` is still showing n the pattern. Did you commit your changes? Would you please edit the pattern again if needed to ensure that it's correct?

Reviewer 2: the `#` is part of the pattern

Author: @pavansokkenagaraj Ah, I read that before having enough coffee this morning. I get it now. Apologies.

For more information about the syntax of the Config custom resource manifest, see [Config](../reference/custom-resource-config) in the _Custom Resources_ section.

## About Regular Expression Validation

> You can use regular expressions (regex) to configure fields that validate user input, ensuring conformity to certain standards, such as valid email addresses, password complexity rules, IP addresses, and URLs. This prevents users from deploying an application with a verifiably invalid configuration and gives them a better user experience.

Reviewer 3: ```suggestion
You can use regular expressions (regex) to validate user input for config items, ensuring conformity to certain standards, such as valid email addresses, password complexity rules, IP addresses, and URLs. This prevents users from deploying an application with a verifiably invalid configuration.
```

> You add the `validation`, `regex`, and `message` fields to the Config custom resource for an item. Regular expressions are supported for `text`, `textarea`, and `password` config option types.

Reviewer 3: ```suggestion
You add the `validation`, `regex`, and `message` fields to items in the Config custom resource. Validation is supported for `text`, `textarea`, and `password` config item types.
```

> The following example shows the requirements for a password:

Reviewer 3: ```suggestion
The following example implements a password complexity rule:
```

```yaml
apiVersion: kots.io/v1beta1
kind: Config
metadata:
  name: my-application
spec:
  groups:
    - name: smtp-settings
      title: SMTP Settings
      items:
      - name: smtp_password
        title: SMTP Password

>         type: text

Reviewer 3: ```suggestion
        type: password
```

        validation:
          regex: ​^.*(?=.{8,20})(?=.*[a-zA-Z])(?=.*\d)(?=.*[!#$%&? "]).*$
          message: Must be a minimum of 8 characters and must contain a combination of uppercase, lowercase, numbers, and special characters.
```

The formatting of the example above is divided into the following parts: 

<table>
    <tr>
      <th width="30%">Regex Pattern</th>
      <th width="70%">Description</th>
    </tr>
    <tr>
      <td>^.*</td>
      <td>(Required) Specifies the start of the regular expression.</td>
    </tr>
    <tr>
      <td>(?=.{8,20})</td>
      <td>Specifies the length of at least 8 characters and a maximum of 20 characters.</td>
    </tr>
    <tr>
      <td>(?=.*[a-zA-Z])</td>
      <td>Specifies that uppercase and lowercase letters are required for the input.</td>
    </tr>
    <tr>
      <td>(?=.*\d)</td>
      <td>Specifies that at least one numerical digit from `0` - `9` is required for the input.</td>
    </tr>
    <tr>
      <td>(?=.*[\W])</td>
      <td>Specifies that at least one special character is required for the input.</td>
    </tr>
    <tr>
      <td>.*$</td>
      <td>(Required) Specifies the end of the regular expression.</td>
    </tr>

>   </table>

Reviewer 3: We should have @pavansokkenagaraj review this. Regular expression syntaxes can vary, so we need to be sure that this matches the syntax that will work in KOTS.

But in my opinion, this isn't necessary either. I don't think we have to put the onus on our docs to teach people how to write a regular expression. It might be better to say what the syntax is that we support (like the regular expression must be recognized by Golang or whatever it is), rather than detailing a particular example like this.

Author: @pavansokkenagaraj Does this example work in KOTS? ^

@ajp-io Agreed, that we don't want to teach them all about regular expressions. But I knew we've had a lot of requests for more examples in our docs in general, and in online forums it seemed that a regular expressions for passwords was confusing for some folks. I was thinking that it might be worthwhile to offer just one password example that work with KOTS. In light of this, do you still think the example might be overkill?

Of course, it does need technical review.

Reviewer 2: `^\w+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$` yes, the standard regex pattern will always work

Reviewer: Only other (unrelated) suggestion might be to move this section after the procedure rather than before it. That way people only need to go look at it if it's something that's relevant to them

Reviewer 3: Cool, I'm fine with keeping this chart as an example then, as long as Pavan confirms that the example is valid and explained correctly.

Author: Thanks, folks.

@pavansokkenagaraj Please confirm that the password complexity example I gave is correct in its coding and in its explanation.

Also, I added the sentence "Standard regular expressions are supported."

Please confirm that these are accurate. Thanks.

Author: Thanks, folks.

@pavansokkenagaraj Please confirm that the password complexity example I gave is correct in its coding and in its explanation.

Also, I added the sentence "Standard regular expressions are supported."

Please confirm that these are accurate. Thanks.

     :::

   * **With the `value` property**: When you include the `value` key, the app manager does not overwrite this value during an application update. The value that you provide for the `value` key is visually indistinguishable from other values that your user provides on the admin console configuration screen. The app manager treats user-supplied values and the value that you provide for the `value` key as the same.

> 1. (Optional) Add regular expressions fields `validation`, `regex`, and `message` to validate  `text`, `textarea`, and `password` config option types. For more information, see [About Regular Expression Validation](#about-regular=expression-validation).

Reviewer 3: ```suggestion
1. (Optional) Add the `validation`, `regex`, and `message` fields to validate user input for `text`, `textarea`, and `password` config item types. For more information, see [About Regular Expression Validation](#about-regular=expression-validation).
```

    **Example**:

      ```yaml
      - name: email_address
        title: Email Address
        type: text
        validation:
          regex: ^\w+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$

>           message: Enter a valid email address.

Reviewer 3: All the other steps here are SMTP related. Should we use the SMTP password example here again?

Author: We could. I was trying to give a different example so they'd have two types. I'll change it to a password example if we end up removing the one above.

Author: @ajp-io @pavansokkenagaraj Not sure if you're going to change the field names that go with `validation`. Please let know if this needs to change.

Reviewer 3: Okay, we'll discuss

Reviewer 2: @ajp-io / @jonquil2002 I have updated the examples to match smtp_password and smtp_hostname
